### ISSUE TITLE
Use `typewriter` device for reifying values as entities

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#116]
A robot *has entities* in its inventory, and also *has values* referred to by its running program.  It's fun to contemplate ways to blur the lines between the two.  This ticket is really meant more to spur discussion than to represent a specific task that needs to be done (perhaps some specific tasks may spin out of it later).

* As mentioned in #115 , we could imagine simulating recursive data structures with nested boxes.  Wouldn't it be fun if a variable holding a list is actually represented as nested `cons` entities in your inventory?  Surely this isn't a great idea in general---it would be way less efficient---but it's fun to think about.
* On the other hand, it's not crazy to imagine wanting to reify, say, a number as a physical inventory entity, so that you can `give` it to another robot, who could reflect it back to an `int` value.  This would be like a poor man's version of #94 before you have developed the necessary tech to properly send values over channels.
* So perhaps we could have a generic mechanism for reifying and reflecting between values and entities.  Would this be interesting? Fun?  Worthwhile?

See also #455 which is probably a prerequisite.

**Comment by Anrock** [comment]
Honestly, ints as physical entities sounds kinda weird. So far the game seems (at least to me) to be sort of realistic, i.e. robot have to have a specific device to do something, with various mechanics tied to devices like scanners. Summoning values into real world feels like cartoonish magic and out of place.

For poor man's versiof of #94 something like "data port" could be done, which is like antenna with 1 tile radius. Which also can be used for #22

**Comment by byorgey** [comment]
> So far the game seems (at least to me) to be sort of realistic

It's not *that* realistic (and I don't intend it to be): there are lambdas and bits you can harvest! And you make branch predictors out of tree branches!  The world is already full of cartoonish, whimsical magic.

**Comment by Anrock** [comment]
Oh, ok! Well then, pile of numbers and throwing ints around definitely fits this setting.

**Comment by byorgey** [comment]
Hmm, how about a "paper" entity and a way to "print" a value onto a piece of paper and then "read" a value from a piece of paper?  Maybe reifying values into physical objects is not so magical after all... =)

**Comment by byorgey** [comment]
So I'm going to make this concrete: how about designing a new `printer` device, which works as follows:

- It enables a new command `print : a -> cmd ()`.  If you have at least one `paper` entity in your inventory, then one is consumed, and you end up with a `printoutNNN` entity (the `NNN` should be chosen to make it unique, I guess?).
- A `scanner` device enables `read : string -> cmd a` which takes the name of a printout entity and reads it back into a value of the expected type (if possible).

Now that I write this down it seems like there are still some issues to be worked out...

**Comment by xsebek** [comment]
I like the printing idea 👍 

Would the expected type be determined from context?
Sounds like it would be necessary to check for escaping type variable like in #351. 😟 

Also what is going to be the “paper” recipe? I would like something like wood+water+furnace so that I could stock up quickly on such useful entity. 🙂

**Comment by byorgey** [comment]
Yes, the expected type would be determined from context, just like Haskell's `show` and `read`.  #351 seems unrelated, I don't foresee any particular issues in terms of the type system.

wood+water+furnace sounds like a good recipe for paper.  I guess the more interesting question to me is what is going to be the recipe for a `printer`?

Some questions I still have:
- There are strange issues with uniqueness of entities etc. related to #114 
- How should you get rid of a `printout` once you are done with it?

**Comment by xsebek** [comment]
About the uniqueness, is there a reason why we can not just modify the entity map?

If we wanted to thwart instant communication via entity numbering, we could make it per robot (or tile). So `"printout #1 by mad_max"`.

Getting rid of it is also interesting, we could add some crusher tool to recycle back to dirt. ♻️ Maybe trees could only be planted on dirt layer.

**Comment by kostmo** [comment]
Note for posterity: there is now a `linotype` device (#1339) to convert a Value to a `string`.

**Comment by byorgey** [comment]
Here's another idea: what if a `typewriter` simply produces an entity with a name like `paper: <value>` where `<value>` is the result of calling `format` on the value?  So, for example, calling `print : a -> Cmd Text` on the value `5` will consume one `paper` entity from your inventory and add a `paper: 5` entity (and return the name of the produced entity).  

Then we can have a `read : Text -> Maybe a` function which is inverse to `format`.  It takes a textual value (optionally prefixed by `paper:`), and returns the corresponding value, if it can be read at the expected type.  So for example `read "5" : Maybe Int = inr 5`, and `read "paper: 5" : Maybe Int = inr 5`, but `read "5" : Maybe Bool = inl ()`.

This way you can `print` a value and `place` the paper somewhere; then another robot can `grab` it and simply `read` the value from the name of the entity.  This way we don't require any special modifications to the existing code to handle entities, and the value printed on some paper would be visible in your inventory.  Entities representing paper with something printed on it will be distinct from "blank" paper entities, and indistinguishable from other paper entities with the same printed value, which seems right.

**Comment by byorgey** [comment]
Closed by #2245.
